### DISCUSSION TITLE
preserveObjectStacking and Editing Text

### DISCUSSION THREAD

**Discussion by milonic** [fabricjs/fabric.js#9824] [Ideas]
Hi Guys,

What do you think about bringing a text object forward if it's behind other objects and preserveObjectStacking is true?

I've just done this in my own code but wondered if it should be in FabricJS as standard practice.

I could code something up if you need me to?

What we thinking?

**Comment by ShaMan123** [comment]
Projects I work on do the same

**Comment by asturur** [comment]
This is why preserveObjectStacking is there.
If you feel like text should be forward when editing occurs, set 'preserveObjectStacking' to false when you select a text.
I don't think there are side effects of changing that at runtime, If there are we should fix eventual bugs.

You have a selection created event, uses that that to switch the behaviour on and off when text is selected. In this way i feel like is already part of fabricJS behaviour because is possible with the api.

**Comment by milonic** [comment]
I definitely tried that but it didn't work for me at the time. Something weird going on in my app because it does work (as you suggested) on a vanilla install of FabricJS.

Next question is: Should it perhaps be a default that editing text is brought forward regardless of preserveObjectStacking flag?

**Reply by asturur** [reply]
No that not, because i really do not want embed high level UX in fabricJS.
The selection methods with shift click and the multi selection with mouse are already too much.

The control api can at least be deactivated and rebuilt from scratch differently by the developer.
I do not think that having strong preferences on how interactivity should work is a good idea

**Reply by milonic** [reply]
Yep, that's fine, I understand.

To be honest, it's quite a trivial thing to do in code. And especially if there's a nice demo and instructions on the new website it will help.

I still can't figure out why my app won't work with preserveObjectStacking  - It's something to do with my implementation of Pointer Events. Which is something else we need to look at but I just need to move forward with what I have for now and address it later.

**Reply by asturur** [reply]
Demos/articles are what saves people time more than many many features.
The issue with 'the default behaviour' is that then you need an option to turn it off and if you don't want to have text coming up when selected or you want it just when editing, then you need more and more boolean configuration switches 'forceTextNoOnTop' you end up doing strange things.
At that point you get that:
- preserveObjectStacking is meant for all objects but not text
- if you want text to preserve the stacking also when editing you have to switch something else

Then there are the mixed selections, what we do?

Things get out of hand very quickly and options start to make less sense.
I regret a lot of options we have
